### ISSUE TITLE
descs: device selection

### ISSUE THREAD

**Issue by sensestage** [ModalityTeam/Modality-toolkit#162]
In order to allow for more specific device definition, I am proposing the following:

`````
// single devices:
( // osc
name: "minibee", // if no device source is indicated, just listen to osc message from anywhere
device: (
	protocol: \osc
	source: ( host: "127.0.0.1", port: 57600 ), // optional; only match indicated host or port
	destination: ( host: "127.0.0.1", port: 57600 ) // optional; port needed if indicated, if no host indicated, use localhost
)
);

// simplest: (just rename device to name)
(
name: "minibee",
protocol: \osc
);


( // midi
name: "LPD8", // use name to find device, unless device source/destination is specified explicitly
device: (
	protocol: \midi,
	source: ( name: "LPD8", port: "LPD8 MIDI 1", portIndex: 0 ), // port index as a fallback for translated midi ports
	destination: ( name: "LPD8", port: "LPD8 MIDI 1", portIndex: 0 )
)
);

// simplest  (just rename device to name)
(
name: "LPD8",
protocol: \midi
);

( // hid
name: "Run'N' Drive_Thrustmaster", // use name to find device, unless device source/destination is specified explicitly
device: (
	protocol: \hid,
	source: ( product: "Run'N' Drive", vendor: "Thrustmaster" ),
	destination: ( product: "Run'N' Drive", vendor: "Thrustmaster" )
);
);

// simplest (just rename device to name)
(
name: "Run'N' Drive_Thrustmaster",
protocol: \hid
);


// multi device:
( // midi + hid
name: "I-Controls",
device: [
		(
			protocol: \midi,
			source: ( name: "I-Control", port: "MIDI 1", portIndex: 0 ),
			destination: ( name: "I-Control", port: "MIDI 1", portIndex: 0 )
		),
		(
			protocol: \hid,
			source: ( product: "I-Control", vendor: "I-Con" )
		)
	]
);

// macky control and midi
(
name: "I-Controls",
device: [
		(
			protocol: \midi,
			source: ( name: "I-Control", port: "MIDI 1", portIndex: 0 ),
			destination: ( name: "I-Control", port: "MIDI 1", portIndex: 0 )
		),
		(
			protocol: \macky,
			source: ( name: "I-Control", port: "MACKY port", vendor: "ICon" ),
			destination: ( name: "I-Control", port: "MACKY port", vendor: "ICon" ),
			features: ( )
		)
	]
);

( // midi + osc
name: "Renoise",
device: (
	[
		(
			protocol: \midi,
			source: ( name: "Renoise", port: "MIDI 1", portIndex: 0 ),
			destination: ( name: "Renoise", port: "MIDI 1", portIndex: 0 )
		),
		(
			protocol: \osc,
			source: ( port: 8000 )
			destination: ( port: 8000 )
		)
	]
);
);
````

**Comment by sensestage** [comment]
```
protocol: ( 
    hid: ( source: ..., destination: ... ),
    midi: ( source: ...., destination: ... )
)
```

**Comment by sensestage** [comment]
instead for multi-devices:

```
devices: // with complete device description
```

**Comment by sensestage** [comment]
```
protocol: \hid,
source:
destination:
device: // if source & destination is the same, as it was
```

**Comment by sensestage** [comment]
for osc now:
( srcPort: ..., ipAddress: ..., recvPort: ..., destPort: ... );
all of these are optional

**Comment by sensestage** [comment]
We renamed the ``device`` to ``idInfo``.

**Comment by sensestage** [comment]
for separate source and destination:
```idInfo: ( source: ( ), destination: () );```

**Comment by sensestage** [comment]
or use multi mktl?
see also #21

**Comment by sensestage** [comment]
For MIDI now this is supported:
```
idInfo: ( name: "Ableton Push", sourcePortIndex: 1, destinationPortIndex: 1 )
```

**Comment by sensestage** [comment]
closed via: 889c316835409042feae318519786fed8f9d9c7b

**Comment by adcxyz** [comment]
would be helpful to have a list of all the actual devices concerned among the devices we currently support.
